src/editor/CodeRefactoringManager.js

```diff
@@ -0,0 +1,231 @@
+     * Providers with a higher number will be asked to refactor code before those
+     * with a lower priority value. Defaults to zero.
+     */
+    function registerRefactoringProvider(providerInfo, languageIds, priority) {
+        var providerObj = { provider: providerInfo,
+            priority: priority || 0 };
+
+        if (languageIds.indexOf("all") !== -1) {
+            // Ignore anything else in languageIds and just register for every language. This includes
+            // the special "all" language since its key is in the refactoringProviders map from the beginning.
+            var languageId;
+            for (languageId in refactoringProviders) {
+                if (refactoringProviders.hasOwnProperty(languageId)) {
+                    refactoringProviders[languageId].push(providerObj);
+                    refactoringProviders[languageId].sort(_providerSort);
+                }
+            }
+        } else {
+            languageIds.forEach(function (languageId) {
+                if (!refactoringProviders[languageId]) {
+                    // Initialize provider list with any existing all-language providers
+                    refactoringProviders[languageId] = Array.prototype.concat(refactoringProviders.all);
+                }
+                refactoringProviders[languageId].push(providerObj);
+                refactoringProviders[languageId].sort(_providerSort);
+            });
+        }
+    }
+
+    /**
+     *  Return the array of refactoring providers for the given language id.
+     *  This gets called on context menu open.
+     *
+     * @param {!string} languageId
+     * @return {?Array.<{provider: Object, priority: number}>}
+     */
+    function _getProvidersForLanguageId(languageId) {
+        var providers = refactoringProviders[languageId] || refactoringProviders.all;
+
+        //@TODO - Exclude providers that are explicitly disabled in the preferences.
+        return providers;
+    }
+
+    /**
+    * This function is responsible for registering command and adding menus for available handlers
+    * Remove submenu in case of no handler found
+    */
+    function _updateHandlers() {
+        var response = refactoringProvider.getHandlers();
+        if (!response) {
+            editorCmenu.removeSubMenu(Commands.CODE_REFACTOR);
+        } else {
+            response.forEach(function(element) {
+                var commandObj = CommandManager.get(element.id);
+                if (!commandObj) {
+                    CommandManager.register(element.name, element.id, element.handler);
+                }
+
+                //Adding Context submenu under Refactor
+                submenu.addMenuItem(element.id);
+            });
+        }
+    }
+
+    /**
+     * Find a suitable provider for a language id and
+     * Add a context menu if there is any provider available
+     */
+    function getHandlersForLanguageId() {
```

Comment:
You might want to change the name of this function as it more than a `getter`. Also all private function need to have `_` prepended.